README.md
## ajanuw-async-validate

Asynchronous form validator

## install
```sh
$ npm i ajanuw-async-validate
```

## Use
```ts
import { AsyncValidate } from "ajanuw-async-validate";

// Create an asynchronous validator
const av = new AsyncValidate(
  {

    name: {
      required: "name is required",
      validators: [
        AsyncValidate.minLength(6, "姓名最少需要6个字符"),

        // Custom validator
        (name:string, data:any) => Promise.resolve( name === 'ajanuw' ? null : { checkName: "name error." } )
      ],

      // When "name" verification fails, this "fail" function will be called
      fail(er) {
        expect(er.errors.minLength).toBe("姓名最少需要6个字符");
        expect(er.errors.checkName).toBeTruthy();
      },
    },

    pwd: {
      required: "密码必填",
      minLength: [8, "密码最少需要8个字符"],
    },

    pwd2: {
      required: "填写确认密码",
      validators: (input, data) => input !== data.pwd ? { checkPwd: "两次密码填写不一样" } : null
      },
    },

  },
  {
    // By default, it will return directly when the first error is detected
    checkAll: true,

    // You can handle all errors here
    // If this "fail" is set, "AsyncValidate.fail" will not be called
    fail(erFields) {
      expect("name" in erFields).toBe(true);
      expect(erFields.name.errors.minLength).toBe("姓名最少需要6个字符");
      expect(erFields.name.errors.checkName).toBeTruthy();

      expect("pwd2" in erFields).toBe(true);
      expect(erFields.pwd2.errors.required).toBeTruthy();
      expect(erFields.pwd2.errors.checkPwd).toBeTruthy();
    },
  }
);

// Verify the data, if all are successful, it will return "true", otherwise it will return "false"

await av.validate({
  name: "aja",
  pwd: "12345678",
  pwd2: "",
}) // false
```


## Set field validator
```ts
// Set up a single
new AsyncValidate({
  name: AsyncValidate.required("name is required!"),
});

// Set multiple
new AsyncValidate({
  name: [ AsyncValidate.required("name is required!") ],
});

// Can also be like this
new AsyncValidate({
  name: {
    required: "name is required!"
  }
})

// Set to "null" to skip the verification of this field
new AsyncValidate({
  name: null,
})


// Verify Object value
const av = new AsyncValidate(
  {
    name: AsyncValidate.required("name is requries!"),
    address: {
      object: "address error.",
      validators: {
          street: {
            required: "require street.",
          },
          zip: {
            required: "require street.",
          },
        },
    },
  },
  {
    fail(erFields) {
      console.error(erFields);
    },
  }
);

await av.validate({
    name: "ajanuw",
    address: {
      street: "",
      zip: "",
    },
})
```


## Use "mixin" to define a global validator
```ts
AsyncValidate.mixin({
  enum(c: any[], msg: string) {
    return (input) => {
      if (!c.includes(input)) return { enum: msg };
    };
  },
});

const av = new AsyncValidate({
  value: {
    enum: [["a", "b", "c"], "error."],
  },
});

await av.validate({ value: "a" }) // true
await av.validate({ value: "d" }) // false
```



## AsyncValidateOptions
```ts
type ValidateHandleArg = string | any[];

export interface AsyncValidateOptions {
  string?: ValidateHandleArg;
  number?: ValidateHandleArg;
  bool?: ValidateHandleArg;
  float?: ValidateHandleArg;
  int?: ValidateHandleArg; // Number.isSafeInteger()
  object?: ValidateHandleArg; // "[object Object]"
  array?: ValidateHandleArg; // Array.isArray()
  json?: ValidateHandleArg; // try JSON.parse()
  email?: ValidateHandleArg;
  hex?: ValidateHandleArg; // 0x0A 0Ah 0A
  regexp?: ValidateHandleArg; // /a/ new RegExp()

  or?: ValidateHandleArg;
  and?: ValidateHandleArg;

  required?: ValidateHandleArg;
  validators?:
    | AsyncValidateHandle
    | AsyncValidateHandle[]
    | IValidateConfig /* to AsyncValidate */;

  // 监听单个字段的错误,当验证失败(invalid)时，调用
  fail?: (errors: { value: any; errors: AnyObject }) => void;

  [name: string]: any;
}
```

## Set error handling for all validators
```ts
import { AsyncValidate } from "ajanuw-async-validate";

AsyncValidate.fail = (erFields) => {
 //...
}
```

## Get the first error string
```ts
AsyncValidate.fail = (erFields) => {
 AsyncValidate.firstError(erFields) // first error string
}
```

## Use on the browser
```html
<script src="ajanuw-async-validate.js"></script>
<script>
  const { AsyncValidate } = window.AjanuwAsyncValidate;

  const av = new AsyncValidate(
    {
      username: {
        required: "请填写用户名!",
        fail(er) {
          // 字段验证失败结果:  ['请填写用户名!']
          console.log("字段验证失败结果: ", Object.values(er.errors));
        },
      },
      password: {
        required: "请填写密码!",
      },
    },
    {
      fail(er) {
        // 表单验证失败结果:  请填写用户名!
        console.log("表单验证失败结果: ", AsyncValidate.firstError(er));
      },
    }
  );

  (async () => {
    const data = {
      username: "",
      password: "",
    };
    const valid = await av.validate(data);
    console.log("valid: ", valid); // false
  })();
</script>
```


## And
```ts
const av = new AV({
  x: {
    required: "必填",
    or: [
      [AV.number(), AV.and([AV.string(), AV.hex()])],
      "必须为数字或则字符串!",
    ],
  },
});

expect(await av.validate({ x: 123 }))   .toBe(true);
expect(await av.validate({ x: "0x01" })).toBe(true);
expect(await av.validate({ x: true }))  .toBe(false);
```

## Or
```ts
const av = new AV({
  x: {
    required: "必填",
    or: [[AV.number(), AV.string()], "必须为数字或则字符串!"],
  },
});

expect(await av.validate({ x: 123 })).toBe(true);
expect(await av.validate({ x: "123" })).toBe(true);
expect(await av.validate({ x: true })).toBe(false);
```

More features, [please see the test](https://github.com/januwA/ajanuw-async-validate/blob/main/test/test.test.ts)

## build
> $ npm run build

## test
> $ npm t

CHANGELOG.md
## 0.9.1 2022-8-1

- 🐛 build new package
  

## 0.9.0 2022-8-1

- ✨ add `len`
- ✅✔️ test `len` and `minLength` and `maxLength`


## 0.8.2 2021-10-27

- 📦 update packages
- 📑 update readme

## 0.8.1 2021-9-20

- 升级依赖
- 浏览器上使用文档

## 0.8.0 2021-8-16

- 增加 `and` `or` 验证器
- msg 错误消息是可选的
- 增加测试

## 0.7.0 2021-8-15

- 修复一些错误，增加测试

## 0.6.0 2021-8-14

- 更新 readme
- oneError 重命名为 firstError


## 0.5.0 2021/1/12

- add "options.ignore"

## 0.4.0 2020/12/28

- update readme
- add `AsyncValidate.oneError`

## 0.3.1 2020/12/1

- update readme
- 优化一些方法

## 0.3.0 2020/11/30

- 大量更新

## 0.2.0 2020/11/27

- 处理错误时，加上name参数

## 0.1.0 2020/11/27

- 一个简单的异步表单验证器

src/async-validate.ts
import {
  AnyObject,
  AsyncValidateHandle,
  ValidateFailFileds,
  ValidateFailHandle,
  ValidateHandleArg,
} from "./interface";

const VALIDATORS = "validators";

export interface AsyncValidateOptions {
  string?: ValidateHandleArg;
  number?: ValidateHandleArg;
  bool?: ValidateHandleArg;
  float?: ValidateHandleArg;
  int?: ValidateHandleArg; // Number.isSafeInteger()
  object?: ValidateHandleArg; // "[object Object]"
  array?: ValidateHandleArg; // Array.isArray()
  json?: ValidateHandleArg; // try JSON.parse()
  email?: ValidateHandleArg;
  hex?: ValidateHandleArg; // 0x0A 0Ah 0A
  regexp?: ValidateHandleArg; // /a/ new RegExp()

  or?: ValidateHandleArg;
  and?: ValidateHandleArg;

  required?: ValidateHandleArg;
  validators?:
    | AsyncValidateHandle
    | AsyncValidateHandle[]
    | IValidateConfig /* to AsyncValidate */;

  // 监听单个字段的错误,当验证失败(invalid)时，调用
  fail?: (errors: { value: any; errors: AnyObject }) => void;

  [name: string]: any;
}

type AsyncValidateCheck =
  | AsyncValidateHandle
  | AsyncValidateHandle[]
  | AsyncValidateOptions;

/**
 * 验证器配置
 */
export interface IValidateConfig {
  [key: string]: null | AsyncValidateCheck;
}

export interface IOptions<T> {
  /**
   * 如果为true会检查完所有字段的验证器，否则检测失败直接返回
   *
   * default: false
   */
  checkAll?: boolean;

  /**
   * 字段验证失败时，用户处理错误的函数
   */
  fail?: ValidateFailHandle<T>;
}

function isSuccess<T>(errorFileds: ValidateFailFileds<T>) {
  return Object.keys(errorFileds).length === 0;
}

function isObject(data: any): boolean {
  return Object.prototype.toString.call(data) === "[object Object]";
}

function hasValidators(obj: any): obj is AsyncValidateOptions {
  return obj.hasOwnProperty("validators");
}

/**
 * 将不同的validators，处理为 AsyncValidateHandle[]
 * @param validators
 */
function handleValidators(
  validators?: AsyncValidateCheck
): AsyncValidateHandle[] {
  if (!validators) return [];

  if (Array.isArray(validators)) return validators;

  if (typeof validators === "function") return [validators];

  if (isObject(validators)) {
    let vs: AsyncValidateHandle[] = [];

    // 映射AsyncValidate上的静态方法，如果不存在直接抛错，应为可能造成错误的结果
    for (const key in validators) {
      if ([VALIDATORS, "fail"].includes(key)) continue;
      if (AsyncValidate.hasOwnProperty(key)) {
        let arg: ValidateHandleArg = (validators as any)[key];
        if (!Array.isArray(arg)) arg = [arg];
        vs.push((AsyncValidate as any)[key](...arg));
      } else {
        throw new Error(`[[ AsyncValidate ]] not "${key}" validate.`);
      }
    }

    if (
      !(validators.validators instanceof AsyncValidate) &&
      !isObject(validators.validators)
    ) {
      vs = vs.concat(handleValidators(validators.validators));
    }

    return vs;
  }

  return [];
}

export class AsyncValidate<
  T extends IValidateConfig,
  D = {
    [key in keyof T]?: any;
  }
> {
  options: IOptions<D>;

  /**
   * 设置所有验证器的错误处理
   */
  static fail?: ValidateFailHandle<any>;

  /**
   * 提取第一个错误字段的第一个错误消息
   *
   * ```js
   * err = AsyncValidate.firstError({
   *   name: { errors: { required: 'name is required!' }, value: ''}
   * })
   *
   * err === 'name is required!'
   * ```
   */
  static firstError(errorFields: ValidateFailFileds<any>) {
    if (errorFields && Object.keys(errorFields).length) {
      return Object.values(errorFields[Object.keys(errorFields)[0]].errors)[0];
    }
  }

  constructor(public readonly validateConfig: T, options?: IOptions<D>) {
    this.options = Object.assign(
      {
        checkAll: false,
      },
      options
    );
  }

  /**
   * 定义全局的验证器
   * 
   * ## Example
   * 
   * ```ts
   AsyncValidate.mixin({
      num(c: any[], msg: string) {
        return (input) => {
          if (!c.includes(input)) return msg;
        };
      },
    })
   * ```
   */
  static mixin(handles: {
    [name: string]: (...args: any[]) => AsyncValidateHandle;
  }) {
    Object.keys(handles)
      .filter((k) => {
        if (!AsyncValidate.hasOwnProperty(k)) return k;
      })
      .forEach((k) => ((AsyncValidate as any)[k] = (handles as any)[k]));
  }

  /**
   *
   * @param keyValidate 验证器
   * @param value 验证的数据
   * @param data 所有数据
   * @param errorCallback 验证失败时的回调
   */
  private async checkValue(
    keyValidate:
      | AsyncValidateHandle
      | AsyncValidateHandle[]
      | AsyncValidateOptions,
    value: any,
    data: AnyObject,
    errorCallback: (validate: AnyObject) => void
  ) {
    for (const h of handleValidators(keyValidate)) {
      const error = await h(value, data);
      if (error) errorCallback(error);
    }
  }

  private _fail(errorFileds: ValidateFailFileds<D>) {
    if (isSuccess(errorFileds)) return;

    // 如果验证器有错误处理，则不调用全局错误处理
    if (this.options.fail) this.options.fail(errorFileds);
    else if (AsyncValidate.fail) AsyncValidate.fail(errorFileds);
  }

  /**
   * 验证数据
   * @param data 需要验证的数据
   * @param handleFail 接收错误回调
   * @returns
   */
  async validate(
    data: D,
    handleFail?: (errorFileds: ValidateFailFileds<D>) => void
  ): Promise<boolean> {
    const errorFileds: ValidateFailFileds<D> = {};
    let success = true;
    data ??= {} as D;

    for (const [key, keyValidate] of Object.entries(this.validateConfig)) {
      // 验证器设置为空，将跳过检测
      if (!keyValidate) continue;

      if (!(key in data)) {
        throw new Error(`AsyncValidate Error: 没有 ${key} 数据!`);
      }

      const value = (data as any)[key];

      await this.checkValue(keyValidate, value, data, (error) => {
        errorFileds[key] ??= {
          value,
          data,
          errors: {},
        };
        Object.assign(errorFileds[key].errors, error);
      });

      // 每个字段中定义的 fail
      if (key in errorFileds && "fail" in keyValidate) {
        keyValidate.fail?.(errorFileds[key]);
      }

      // 验证失败调用fail
      success = isSuccess(errorFileds);
      if (!success && !this.options.checkAll) break;

      if (hasValidators(keyValidate) && isObject(keyValidate.validators)) {
        const av = new AsyncValidate<any, any>(
          keyValidate.validators as IValidateConfig,
          this.options
        );
        success = await av.validate(value);
        if (!success && !this.options.checkAll) break;
      }
    } // for end

    if (!success) {
      this._fail(errorFileds);
      handleFail?.(errorFileds);
    }
    return success;
  }

  //======================👇静态验证函数========================//

  /**
   *
   * @param validators 测试表达式
   * @param msg 错误消息
   * @returns
   */
  static and(validators: AsyncValidateHandle[], msg?: string) {
    return async (input: any, data: any) => {
      for await (const v of validators) {
        if (await v(input, data)) return { and: msg };
      }

      return null;
    };
  }

  /**
   *
   * @param validators 测试表达式
   * @param msg 错误消息
   * @returns
   */
  static or(validators: AsyncValidateHandle[], msg?: string) {
    return async (input: any, data: any) => {
      let r: any;
      for await (const v of validators) {
        r = await v(input, data);
        if (!r) return null;
      }

      return { or: msg };
    };
  }

  // 必填
  static required(msg?: string): AsyncValidateHandle {
    return (input) => {
      if (!input) return { required: msg };
    };
  }

  // length 长度判断
  static len(size: number, msg?: string): AsyncValidateHandle {
    return (input) => {
      if (!input.hasOwnProperty("length") || input.length !== size)
        return { len: msg };
    };
  }

  // 最小长度
  static minLength(len: number, msg?: string): AsyncValidateHandle {
    return (input) => {
      if (
        !input.hasOwnProperty("length") ||
        input.length < parseFloat(len as any)
      )
        return { minLength: msg };
    };
  }

  // 最大长度
  static maxLength(len: number, msg?: string): AsyncValidateHandle {
    return (input) => {
      if (
        !input.hasOwnProperty("length") ||
        input.length > parseFloat(len as any)
      )
        return { maxLength: msg };
    };
  }

  // 简单的验证手机号
  private static PHONE_EXP =
    /^((13[0-9])|(14[0-9])|(15[0-9])|(16[0-9])|(17[0-9])|(18[0-9])|(19[0-9]))\d{8}$/;
  static phone(msg?: string): AsyncValidateHandle {
    return (input) => {
      if (typeof input === "string" && !input.match(AsyncValidate.PHONE_EXP)) {
        return { phone: msg };
      }
    };
  }

  // 简单的判断相等
  static eql(data: any, msg?: string): AsyncValidateHandle {
    return (input) => {
      if (input !== data) return { eql: msg };
    };
  }

  // 简单的判断相等
  static equal(data: any, msg?: string): AsyncValidateHandle {
    return (input) => {
      if (input != data) return { equal: msg };
    };
  }

  /**
   * 必须为bool
   * @param msg
   */
  static bool(msg?: string): AsyncValidateHandle {
    return (input) => {
      if (!(input === true || input === false)) return { bool: msg };
    };
  }

  private static EMAIL_EXP =
    /^(?=.{1,254}$)(?=.{1,64}@)[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+)*@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;

  static email(msg?: string): AsyncValidateHandle {
    return (input: string) => {
      if (!input.match(AsyncValidate.EMAIL_EXP)) return { email: msg };
    };
  }

  static max(max: number, msg?: string): AsyncValidateHandle {
    return (input) => {
      if (!isNaN(input) && input > parseFloat(max as any)) return { max: msg };
    };
  }

  static min(min: number, msg?: string): AsyncValidateHandle {
    return (input) => {
      if (!isNaN(input) && input < parseFloat(min as any)) return { min: msg };
    };
  }

  static hex(msg?: string): AsyncValidateHandle {
    return (input: string) => {
      input = input.replace(/^0x/i, "");
      input = input.replace(/h$/i, "");
      if (/[^0-9a-fA-F]/.test(input)) return { hex: msg };
    };
  }

  static number(msg?: string): AsyncValidateHandle {
    return (input: number) => {
      // 必须是有穷数的数字
      if (!Number.isFinite(input) || typeof input !== "number")
        return { number: msg };
    };
  }

  static int(msg?: string): AsyncValidateHandle {
    return (input: number, data) => {
      if (this.number("")(input, data) || !Number.isSafeInteger(input))
        return { int: msg };
    };
  }

  static float(msg?: string): AsyncValidateHandle {
    return (input: number, data) => {
      if (this.number("")(input, data) || Number.isInteger(input))
        return { float: msg };
    };
  }

  static array(msg?: string): AsyncValidateHandle {
    return (input: number) => {
      if (!Array.isArray(input)) return { array: msg };
    };
  }

  static object(msg?: string): AsyncValidateHandle {
    return (input) => {
      if (Object.prototype.toString.call(input) !== "[object Object]")
        return { object: msg };
    };
  }

  static json(msg?: string): AsyncValidateHandle {
    return (input) => {
      try {
        JSON.parse(input);
      } catch (error) {
        return { json: msg };
      }
    };
  }

  static regexp(msg?: string): AsyncValidateHandle {
    return (input) => {
      if (!(input instanceof RegExp)) return { regexp: msg };
    };
  }

  static string(msg?: string): AsyncValidateHandle {
    return (input) => {
      if (typeof input === "string") return null;
      return { string: msg };
    };
  }
}
